--- ./misc/Makefile.in.orig	2021-12-04 06:49:39.172274642 +0000
+++ ./misc/Makefile.in	2021-12-04 06:50:30.782883475 +0000
@@ -227,7 +227,7 @@
 
 e2image.static: $(E2IMAGE_OBJS) $(PROFILED_DEPLIBS) $(DEPLIBBLKID)
 	$(E) "	LD $@"
-	$(Q) $(CC) $(LDFLAGS_STATIC) -g -pg -o e2image.static \
+	$(Q) $(CC) $(LDFLAGS_STATIC) -o e2image.static \
 		$(E2IMAGE_OBJS) $(STATIC_LIBS) $(LIBINTL) $(SYSLIBS) \
 		$(STATIC_LIBBLKID) $(LIBMAGIC)
 
