--- /var/db/repos/gentoo/media-libs/mesa/mesa-21.3.0.ebuild	2021-11-20 06:19:38.938285080 -0600
+++ mesa-21.3.0.ebuild	2021-11-20 06:35:00.327072986 -0600
@@ -27,7 +27,7 @@
 "
 
 RADEON_CARDS="r100 r200 r300 r600 radeon radeonsi"
-VIDEO_CARDS="${RADEON_CARDS} crocus freedreno i915 i965 intel iris lima nouveau panfrost v3d vc4 virgl vivante vmware"
+VIDEO_CARDS="${RADEON_CARDS} d3d12 crocus freedreno i915 i965 intel iris lima nouveau panfrost v3d vc4 virgl vivante vmware"
 for card in ${VIDEO_CARDS}; do
 	IUSE_VIDEO_CARDS+=" video_cards_${card}"
 done
@@ -43,6 +43,7 @@
 	vulkan? ( video_cards_radeonsi? ( llvm ) )
 	vulkan-overlay? ( vulkan )
 	video_cards_crocus? ( gallium )
+	video_cards_d3d12?      ( gallium )
 	video_cards_freedreno?  ( gallium )
 	video_cards_intel?  ( classic )
 	video_cards_i915?   ( || ( classic gallium ) )
@@ -217,6 +218,7 @@
 	opencl? (
 		>=sys-devel/gcc-4.6
 	)
+	video_cards_d3d12? ( media-libs/directx-headers )
 	sys-devel/bison
 	sys-devel/flex
 	virtual/pkgconfig
@@ -440,7 +442,7 @@
 		   use video_cards_vivante; then
 			gallium_enable -- kmsro
 		fi
-
+		gallium_enable video_cards_d3d12 d3d12
 		gallium_enable -- swrast
 		gallium_enable video_cards_lima lima
 		gallium_enable video_cards_panfrost panfrost
